app/waf/src/controllers/application.controller.ts

```diff
@@ -299,96 +437,50 @@ export class ApplicationController {
     responses: {
       '204': Schema.emptyResponse('Successfully deploy Application resource'),
       '404': Schema.notFound('Can not find Application resource'),
+      '422': Schema.unprocessableEntity('Fail to deploy Application resource'),
     },
   })
   async deployById(
     @param(Schema.pathParameter('applicationId', 'Application resource ID'))
     id: string,
-  ): Promise<Object> {
+  ): Promise<string> {
     let application = await this.applicationRepository.findById(id);
 
-    let services = await this.applicationRepository
-      .services(application.id)
-      .find();
-    if (services.length === 0) {
+    if (!application.adcId || !application.currentDeclarationId) {
       throw new HttpErrors.UnprocessableEntity(
-        'No service in Application ' + application.id,
+        'No target ADC or no Application declaration to perform deploy action',
       );
     }
 
-    let params: {[key: string]: Object} = {
-      application: application,
-    };
-
-    let tenantAssocs = await this.tenantAssociationRepository.find({
-      limit: 1,
-      where: {
-        tenantId: application.tenantId,
-      },
-    });
-
-    if (tenantAssocs.length === 0) {
-      tenantAssocs = await this.tenantAssociationRepository.find({
-        limit: 1,
-        where: {
-          tenantId: 'default',
-        },
-      });
+    let adc = await this.adcRepository.findById(application.adcId);
+    let declaration = await this.declarationRepository.findById(
+      application.currentDeclarationId,
+    );
 
-      if (tenantAssocs.length === 0) {
-        throw new HttpErrors.UnprocessableEntity(
-          'No ADC associated with Application ' + application.id,
-        );
-      }
-    }
+    let req = new AS3DeployRequest(adc, application, declaration);
+    return await this.as3Service.deploy(AS3_HOST, AS3_PORT, req);
+  }
 
-    let adcs = await this.adcRepository.find({
-      where: {
-        id: tenantAssocs[0].adcId,
-      },
-    });
+  @post(prefix + '/applications/{applicationId}/cleanup', {
+    responses: {
+      '204': Schema.emptyResponse('Successfully cleanup Application resource'),
+      '404': Schema.notFound('Can not find Application resource'),
+    },
+  })
+  async cleanupById(
```

Comment:
As to the API spec, I prefer
```
<obj>/{objId}/action:
body: {
    <actionName>: {parameters}
}
```

As we talked before, following OpenStack endpoint standard.

Reply:
We have not yet made a decision about resource action API style. OpenStack resource action API is not very good, because it will have us to create some extra coding (eg. type definition), just like what you did for ADC resource. Personally, I will prefer a generic way, for example:

```
POST /resource-url-path/action
{
  "action": "deploy",
  "paramaters": {
    "a": 1,
    "b": 2
  }
}
```

The action response can be 204, resource content, or a specific response body. In this way, we will not need to define every action type in code. What do you think?

Reply:
And another alternative is to declare the action name in path, as what I did in this PR.

```
POST /resource-url-path/{action_name}
{body is optional}
```

Its advantage is that it will be flexible for us to define the request/response schema for every action APIs. Our previous two options both register action handers with one fixed url path. That means we have to define the same request/response schema for all action APIs. It will be a little hard to generate very detail action API document.